### PR TITLE
[tripy] Batchnorm implementation

### PR DESCRIPTION
**Pull request by farazkh80**
Batchnorm implementation, currently be used in ResNet50 notebook PR, adding it as a general module similar to `LayerNorm`.

Example
```python
batch_norm = tp.BatchNorm(3)

input = tp.iota((2, 3, 1, 1))
output = batch_norm(input)
```

Tested the code and documnation.

### CODE REVIEW

File: tripy/tripy/frontend/module/batchnorm.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+
+from tripy import export, utils
+from tripy.common import datatype
+from tripy.frontend.module.module import Module
+from tripy.frontend.module.parameter import DefaultParameter, Parameter
+
+@export.public_api(document_under="operations/modules")
+@dataclass
+@utils.constant_fields(["num_features"])
+class BatchNorm(Module):
```
Comment (pranavm-nvidia) [inline tripy/tripy/frontend/module/batchnorm.py:28]: Can you add tests under `tests/integration`? You can refer to `tripy/tests/integration/test_layernorm.py` as an example

File: tripy/tripy/frontend/module/batchnorm.py [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+    The mean (:math:`\mu`) and variance (:math:`\sigma^2`) used are the precomputed running statistics.
+
+    :math:`y = \frac{x - \mu}{\sqrt{\sigma^2 + \epsilon}} * \gamma + \beta`
+
+    where:
+        - :math:`\mu` is the precomputed running mean.
+        - :math:`\sigma^2` is the precomputed running variance.
+        - :math:`\gamma` and :math:`\beta` are learnable parameter vectors (wieight and bias).
+
+    This module is designed for evaluation purposes only, and it does not compute batch statistics.
```
Comment (pranavm-nvidia) [inline tripy/tripy/frontend/module/batchnorm.py:43]: Tripy is inference-only so I think this part is implied. 
```suggestion
```

File: tripy/tripy/frontend/module/batchnorm.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+    This implementation supports 1D, 2D, and 3D inputs (e.g., time-series, images, and volumetric data).
+    Batch Normalization normalizes across the specified feature dimension (typically the second dimension in the input).
+    The mean (:math:`\mu`) and variance (:math:`\sigma^2`) used are the precomputed running statistics.
+
+    :math:`y = \frac{x - \mu}{\sqrt{\sigma^2 + \epsilon}} * \gamma + \beta`
+
+    where:
+        - :math:`\mu` is the precomputed running mean.
+        - :math:`\sigma^2` is the precomputed running variance.
+        - :math:`\gamma` and :math:`\beta` are learnable parameter vectors (wieight and bias).
```
Comment (pranavm-nvidia) [inline tripy/tripy/frontend/module/batchnorm.py:41]: I would slightly shuffle this around. I think we can also drop the part explaining where mean/variance come from (because they come from the user):
```suggestion
    Applies batch normalization over an N-dimensional input tensor using precomputed statistics:
    
    :math:`y = \frac{x - \mu}{\sqrt{\sigma^2 + \epsilon}} * \gamma + \beta`

    where:
        - :math:`\mu` is the precomputed running mean.
        - :math:`\sigma^2` is the precomputed running variance.
        - :math:`\gamma` and :math:`\beta` are learnable parameter vectors (wieight and bias).

    This implementation supports 1D, 2D, and 3D inputs (e.g., time-series, images, and volumetric data).
    Batch Normalization normalizes across the specified feature dimension (typically the second dimension in the input).
```

File: tripy/tripy/frontend/module/batchnorm.py [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+    """
+
+    num_features: int
+    r"""The number of feature channels in the input tensor (the size of the second dimension)."""
+
+    weight: Parameter
+    r"""The learnable scale parameter (γ) of shape `(num_features,)`."""
+
+    bias: Parameter
+    r"""The learnable shift parameter (β) of shape `(num_features,)`."""
```
Comment (pranavm-nvidia) [inline tripy/tripy/frontend/module/batchnorm.py:53]: These parameters are already learned at this point, so I'd drop the "learnable" part. Also, let's stick to ASCII characters as much as possible:
```suggestion
    r"""The :math:`\gamma` parameter of shape :math:`[\text{num_features}]`."""

    bias: Parameter
    r"""The :math:`\beta` parameter of shape :math:`[\text{num_features}]`."""
```

File: tripy/tripy/frontend/module/batchnorm.py [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+
+    running_mean: Parameter
+    r"""The precomputed running mean for the feature channels of shape `(num_features,)`."""
+
+    running_var: Parameter
+    r"""The precomputed running variance for the feature channels of shape `(num_features,)`."""
```
Comment (pranavm-nvidia) [inline tripy/tripy/frontend/module/batchnorm.py:59]: Everything here is precomputed, so we can omit that:
```suggestion
    r"""The running mean for the feature channels of shape :math:`[\text{num_features}]`."""

    running_var: Parameter
    r"""The running variance for the feature channels of shape :math:`[\text{num_features}]`."""
```

File: tripy/tripy/frontend/module/batchnorm.py [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+            input = tp.iota((2, 3, 1, 1))
+            output = batch_norm(input)
+        """
+        super().__init__()
+
+        self.num_features = num_features
+        self.eps = eps
+
+        # Initialize learnable parameters (scale and shift)
+        self.weight = DefaultParameter((num_features,), dtype=datatype.float32)
```
Comment (pranavm-nvidia) [inline tripy/tripy/frontend/module/batchnorm.py:86]: Could we pass the data type as a parameter?

File: tripy/tripy/frontend/module/batchnorm.py [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+    where:
+        - :math:`\mu` is the precomputed running mean.
+        - :math:`\sigma^2` is the precomputed running variance.
+        - :math:`\gamma` and :math:`\beta` are learnable parameter vectors (wieight and bias).
+
+    This module is designed for evaluation purposes only, and it does not compute batch statistics.
+    """
+
+    num_features: int
+    r"""The number of feature channels in the input tensor (the size of the second dimension)."""
```
Comment (pranavm-nvidia) [inline tripy/tripy/frontend/module/batchnorm.py:47]: Should we just call it "channels" instead of "features"?

File: tripy/tripy/frontend/module/batchnorm.py [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+        self.running_mean = DefaultParameter((num_features,), dtype=datatype.float32)
+        self.running_var = DefaultParameter((num_features,), dtype=datatype.float32)
+
+    def __call__(self, x: "tripy.Tensor") -> "tripy.Tensor":
+        r"""
+        Args:
+            x: The input tensor with shape :math:`(N, C, ...)`, where C is the feature dimension.
+
+        Returns:
+            A tensor of the same shape as the input, with batch normalization applied using the precomputed running mean and variance.
```
Comment (pranavm-nvidia) [inline tripy/tripy/frontend/module/batchnorm.py:99]: "with batch norm applied" can be assumed IMO
```suggestion
            A tensor of the same shape as the input.
```

File: tripy/tripy/frontend/module/batchnorm.py [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+        """
+        from tripy.frontend.trace.ops.unary_elementwise import rsqrt
+        from tripy.frontend.trace.ops.reshape import reshape
+
+        x_shape = (1, -1, *([1] * (len(x.shape) - 2)))
```
Comment (pranavm-nvidia) [inline tripy/tripy/frontend/module/batchnorm.py:104]: We should use `x.rank` so that we don't need to evaluate `x.shape` here.

File: tripy/tripy/frontend/module/batchnorm.py (hunk 6 above)
Comment (farazkh80) [reply tripy/tripy/frontend/module/batchnorm.py:84]: done

File: tripy/tripy/frontend/module/batchnorm.py (hunk 7 above)
Comment (farazkh80) [reply tripy/tripy/frontend/module/batchnorm.py:45]: I followed torch like naming, but happy to change it to `channels` if you see fit

File: tripy/tripy/frontend/module/batchnorm.py (hunk 9 above)
Comment (farazkh80) [reply tripy/tripy/frontend/module/batchnorm.py:102]: done

File: tripy/tripy/frontend/module/batchnorm.py (hunk 1 above)
Comment (farazkh80) [reply tripy/tripy/frontend/module/batchnorm.py:28]: I added a test against `torch.BatchNorm2D`. Let me know if you want tests `against` 1D and 3D ones as well, but I think their functionally the same as 2D one.

File: tripy/tests/integration/test_batchnorm.py [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+import re
+
+import tripy as tp
+from tests import helper
+
+DTYPES = [(torch.float32, tp.float32), (torch.float16, tp.float16)]
+
+class TestBatchNorm:
+
+    @pytest.mark.l1
```
Comment (pranavm-nvidia) [inline tripy/tests/integration/test_batchnorm.py:12]: How slow are these tests?

File: tripy/tripy/frontend/module/batchnorm.py [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+        Args:
+            x: The input tensor with shape :math:`(N, C, ...)`, where C is the feature dimension.
+
+        Returns:
+            A tensor of the same shape as the input.
+        """
+        from tripy.frontend.trace.ops.unary_elementwise import rsqrt
+        from tripy.frontend.trace.ops.reshape import reshape
+
+        x_shape = (1, -1, *([1] * (x.rank - 2)))
```
Comment (pranavm-nvidia) [inline tripy/tripy/frontend/module/batchnorm.py:105]: Since we know the dimension, can we just use `self.num_features` instead of `-1`?

File: tripy/tests/integration/test_batchnorm.py (hunk 10 above)
Comment (farazkh80) [reply tripy/tests/integration/test_batchnorm.py:12]: they are like 10-12 seconds, should I remove `l1`?

Comment (pranavm-nvidia) [reply tripy/tests/integration/test_batchnorm.py:12]: Can you try doing the reference computation and comparison on GPU? The device -> host copy adds unnecessary overhead.

File: tripy/tripy/frontend/module/batchnorm.py (hunk 11 above)
Comment (farazkh80) [reply tripy/tripy/frontend/module/batchnorm.py]: done

File: tripy/tests/integration/test_batchnorm.py (hunk 10 above)
Comment (farazkh80) [reply tripy/tests/integration/test_batchnorm.py]: done, and removed L1

**Review by pranavm-nvidia — APPROVED** [review]

**Comment by farazkh80** [comment]
closing since I have feature branch PR now
